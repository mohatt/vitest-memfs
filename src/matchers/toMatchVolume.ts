import { Volume } from 'memfs'
import { compareVolumeMaps, createMatcher, VolumeCompareListMatch, volumeToMap } from '@/utils'

interface VolumeMatcherOptions {
  listMatch?: VolumeCompareListMatch
}

declare module 'vitest' {
  interface Matchers<T = any> {
    /**
     * Compare this memfs volume against another memfs volume.
     */
    toMatchVolume(expected: Volume, options?: VolumeMatcherOptions): T
  }
}

const matcherName = 'toMatchVolume'

export default createMatcher(matcherName, function (received, expected, options) {
  const { utils } = this
  if (!(received instanceof Volume)) {
    return {
      pass: false,
      message: () => `expected ${utils.printReceived(received)} to be a memfs Volume instance`,
      actual: received,
      expected: new (class Volume {})(),
    }
  }

  if (!(expected instanceof Volume)) {
    throw new TypeError(
      `You must provide a memfs Volume instance to ${utils.matcherHint(matcherName)}, not '${typeof expected}'.`,
    )
  }

  if (received === expected) {
    return {
      pass: true,
      message: () => 'volumes matched by reference',
    }
  }

  const receivedMap = volumeToMap(received)
  const expectedMap = volumeToMap(expected)

  return compareVolumeMaps(receivedMap, expectedMap, options)
})
