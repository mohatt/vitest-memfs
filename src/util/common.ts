import { vi, expect } from 'vitest'

/**
 * Imports the real `fs/promises` module, bypassing Vitest mocks.
 *
 * A direct import works in most cases, but if `fs` is mocked
 * in a Vitest setup file it will return the mock instead.
 */
export async function importActualFS() {
  return vi.importActual<typeof import('fs/promises')>('fs/promises')
}

export function isPlainObject(value: unknown): value is Record<string, any> {
  return Object.prototype.toString.call(value) === '[object Object]'
}

type MatchersObject = Parameters<(typeof expect)['extend']>[0]

/**
 * Creates and returns a matcher function.
 * We could wrap the matcher function in the future to warn on un-awaited promises.
 */
export function createMatcher<T extends keyof MatchersObject>(matcher: T, fn: MatchersObject[T]) {
  Object.defineProperty(fn, 'name', { value: matcher })
  return fn
}
