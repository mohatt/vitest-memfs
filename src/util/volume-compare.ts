import { createHash } from 'node:crypto'
import { isText } from 'istextorbinary'
import type { VolumeMap, VolumeMapEntry } from './volume.js'

export type VolumeCompareListMatch = 'exact' | 'ignore-extra' | 'ignore-missing'
export type VolumeCompareReportType = 'first' | 'all'

export interface VolumeCompareOptions {
  listMatch?: VolumeCompareListMatch
  report?: VolumeCompareReportType
}

export interface VolumeCompareResult {
  pass: boolean
  message: () => string
  actual?: unknown
  expected?: unknown
}

/**
 * Compare two volume maps.
 *
 * @param received map from volume under test
 * @param expected map from reference (snapshot or other volume)
 * @param options extra options
 */
export function compareVolumeMaps(
  received: VolumeMap,
  expected: VolumeMap,
  options?: VolumeCompareOptions,
): VolumeCompareResult {
  if (options?.report === 'all') {
    return compareVolumeMapsFull(received, expected, options)
  }

  const listMatch = options?.listMatch
  // make sorted arrays for error reporting and better diffing
  const actualFiles = Object.keys(received).sort()
  const expectedFiles = Object.keys(expected).sort()

  let listMismatchResult: {
    reason: string
    actual: string[]
    expected: string[]
  } | null = null

  switch (listMatch) {
    case 'ignore-extra': {
      const missing = expectedFiles.filter((f) => !(f in received))
      if (missing.length > 0) {
        listMismatchResult = {
          reason: `volume is missing ${missing.length} expected file${missing.length > 1 ? 's' : ''}`,
          actual: actualFiles.filter((f) => f in expected),
          expected: expectedFiles,
        }
      }
      break
    }

    case 'ignore-missing': {
      const extra = actualFiles.filter((f) => !(f in expected))
      if (extra.length > 0) {
        listMismatchResult = {
          reason: `volume has ${extra.length} unexpected file${extra.length > 1 ? 's' : ''}`,
          actual: actualFiles,
          expected: expectedFiles.filter((f) => f in received),
        }
      }
      break
    }

    case 'exact':
    default: {
      if (
        actualFiles.length !== expectedFiles.length ||
        !actualFiles.every((f, i) => f === expectedFiles[i])
      ) {
        listMismatchResult = {
          reason: 'directory structure didnâ€™t match',
          actual: actualFiles,
          expected: expectedFiles,
        }
      }
    }
  }

  if (listMismatchResult) {
    return {
      pass: false,
      message: () => listMismatchResult.reason,
      actual: listMismatchResult.actual,
      expected: listMismatchResult.expected,
    }
  }

  // compare contents
  const filesToCheck = listMatch === 'ignore-missing' ? actualFiles : expectedFiles
  for (const file of filesToCheck) {
    const exp = expected[file]
    const act = received[file]
    if (exp.type !== act.type) {
      return {
        pass: false,
        message: () => `path type mismatch at \`${file}\``,
        actual: makeEntryPreview(file, act),
        expected: makeEntryPreview(file, exp),
      }
    }
    if (exp.type === 'file') {
      const expBuff = exp.data
      const actBuff = (act as typeof exp).data
      if (!expBuff.equals(actBuff)) {
        return {
          pass: false,
          message: () => `mismatch in file \`${file}\``,
          actual: makeEntryPreview(file, act),
          expected: makeEntryPreview(file, exp),
        }
      }
    } else if (exp.type === 'symlink') {
      const expTarget = exp.target
      const actTarget = (act as typeof exp).target

      if (expTarget !== actTarget) {
        return {
          pass: false,
          message: () => `symlink target mismatch at \`${file}\``,
          actual: makeEntryPreview(file, act),
          expected: makeEntryPreview(file, exp),
        }
      }
    }
  }

  // everything is matched at this point
  return {
    pass: true,
    message: () => 'volumes matched',
  }
}

export function compareVolumeMapsFull(
  received: VolumeMap,
  expected: VolumeMap,
  options?: Omit<VolumeCompareOptions, 'report'>,
) {
  const listMatch = options?.listMatch
  const actualDiff: Record<string, unknown> = {}
  const expectedDiff: Record<string, unknown> = {}
  let missingCount = 0
  let extraCount = 0
  let diffCount = 0

  function addDiff(f: string, exp?: VolumeMapEntry, act?: VolumeMapEntry) {
    if (act) actualDiff[f] = makeEntryPreview(f, act)
    if (exp) expectedDiff[f] = makeEntryPreview(f, exp)
  }

  function addMatch(f: string) {
    actualDiff[f] = {}
    expectedDiff[f] = {}
  }

  const filesToCheck =
    listMatch === 'ignore-extra'
      ? expected
      : listMatch === 'ignore-missing'
        ? received
        : { ...received, ...expected }
  for (const f in filesToCheck) {
    const exp = expected[f]
    const act = received[f]

    if (exp && act) {
      if (exp.type !== act.type) {
        addDiff(f, exp, act)
        diffCount++
      } else if (exp.type === 'file') {
        if (!exp.data.equals((act as typeof exp).data)) {
          addDiff(f, exp, act)
          diffCount++
        } else {
          addMatch(f)
        }
      } else if (exp.type === 'symlink') {
        if (exp.target !== (act as typeof exp).target) {
          addDiff(f, exp, act)
          diffCount++
        } else {
          addMatch(f)
        }
      } else {
        addMatch(f)
      }
    } else if (exp && !act && listMatch !== 'ignore-missing') {
      addDiff(f, exp)
      missingCount++
    } else if (act && !exp && listMatch !== 'ignore-extra') {
      addDiff(f, null, act)
      extraCount++
    }
  }

  const total = missingCount + extraCount + diffCount
  if (total > 0) {
    const parts: string[] = []
    if (missingCount) parts.push(`${missingCount} missing path${missingCount > 1 ? 's' : ''}`)
    if (extraCount) parts.push(`${extraCount} unexpected path${extraCount > 1 ? 's' : ''}`)
    if (diffCount) parts.push(`${diffCount} mismatched content`)

    return {
      pass: false,
      message: () =>
        parts.length === 1
          ? `Found ${parts[0]}` //
          : `Found ${total} mismatches: ${parts.join(', ')}`,
      actual: actualDiff,
      expected: expectedDiff,
    }
  }

  return { pass: true, message: () => 'volumes matched' }
}

function makeEntryPreview(name: string, entry: VolumeMapEntry) {
  if (entry.type === 'empty-dir') {
    return new Directory()
  }
  if (entry.type === 'file') {
    if (isText(name, entry.data)) {
      return new TextFile(entry.data)
    }
    return new BinaryFile(entry.data)
  }
  return new Symlink(entry.target)
}

class TextFile {
  data: string
  constructor(buff: Buffer) {
    this.data = buff.toString('utf8')
  }
}

class BinaryFile {
  hash: string
  length: number
  preview: string
  constructor(buff: Buffer, trim = 32) {
    this.hash = createHash('sha1').update(buff).digest('hex')
    this.length = buff.length
    const head = buff.subarray(0, trim).toString('base64')
    const tail = buff.subarray(buff.length - trim).toString('base64')
    this.preview = `${head}...${tail}`
  }
}

class Symlink {
  constructor(public target: string) {}
}

class Directory {}
