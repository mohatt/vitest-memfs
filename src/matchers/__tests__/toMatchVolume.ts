import { describe, it, expect } from 'vitest'
import { Volume, vol } from 'memfs'
import { makeTests, makeVol, VolumeInput } from '@test/util.js'
import toMatchVolume, { VolumeMatcherOptions } from '../toMatchVolume.js'

interface TestCase {
  name: string
  left: VolumeInput
  right: VolumeInput
  opts?: VolumeMatcherOptions
  pass?: boolean
  not?: boolean
}

const cases = makeTests<TestCase>([
  {
    name: 'identical files',
    left: { '/foo.txt': 'hi' },
    right: { '/foo.txt': 'hi' },
    pass: true,
  },
  {
    name: 'same ref (singleton vol)',
    left: () => {
      vol.reset()
      vol.fromJSON({ '/foo.txt': 'hi' })
      return vol
    },
    right: () => vol,
    pass: true,
  },
  {
    name: 'content mismatch',
    left: { '/foo.txt': 'hello' },
    right: { '/foo.txt': 'world' },
    pass: false,
  },
  {
    name: 'accepts json input',
    left: { '/foo.txt': 'hi' },
    right: () => ({ '/foo.txt': 'hi' }) as any,
    pass: true,
  },
  {
    name: 'invalid type (received)',
    left: () => 'invalid' as any,
    right: () => vol,
  },
  {
    name: 'invalid type (expected)',
    left: () => vol,
    right: () => 'invalid' as any,
  },
  {
    name: 'invalid object type (expected)',
    left: () => vol,
    right: () => new Date() as any,
  },
  {
    name: 'missing file',
    left: { '/foo.txt': 'hi' },
    right: { '/foo.txt': 'hi', '/bar.txt': 'extra' },
    pass: false,
  },
  {
    name: 'extra file',
    left: { '/foo.txt': 'hi', '/bar.txt': 'extra' },
    right: { '/foo.txt': 'hi' },
    pass: false,
  },
  {
    name: 'respects ignore-extra option',
    left: { '/foo.txt': 'hi', '/bar.txt': 'hey', '/extra.txt': 'extra' },
    right: { '/foo.txt': 'hi', '/bar.txt': 'hey' },
    pass: true,
    opts: { listMatch: 'ignore-extra' },
  },
  {
    name: 'respects ignore-extra option (mismatch)',
    left: { '/foo.txt': 'hi', '/extra.txt': 'extra' },
    right: { '/foo.txt': 'hi', '/bar.txt': 'hey' },
    pass: false,
    opts: { listMatch: 'ignore-extra' },
  },
  {
    name: 'respects ignore-missing option',
    left: { '/foo.txt': 'hi' },
    right: { '/foo.txt': 'hi', '/bar.txt': 'hey' },
    pass: true,
    opts: { listMatch: 'ignore-missing' },
  },
  {
    name: 'respects ignore-missing option (mismatch)',
    left: { '/foo.txt': 'hi', '/extra.txt': 'extra' },
    right: { '/foo.txt': 'hi', '/bar.txt': 'hey' },
    pass: false,
    opts: { listMatch: 'ignore-missing' },
  },
  {
    name: 'empty directories match',
    left: { '/empty': null },
    right: { '/empty': null },
    pass: true,
  },
  {
    name: 'empty dir vs missing dir',
    left: { '/empty': null },
    right: {},
    pass: false,
  },
  {
    name: 'symlink target mismatch',
    left: () => {
      const v = new Volume()
      v.symlinkSync('/target1.txt', '/link.txt')
      return v
    },
    right: () => {
      const v = new Volume()
      v.symlinkSync('/target2.txt', '/link.txt')
      return v
    },
    pass: false,
  },
  {
    name: 'respects prefix option',
    left: { '/src/foo.txt': 'hi' },
    right: { '/src/foo.txt': 'hi', '/bar.txt': 'extra' },
    pass: true,
    opts: { prefix: '/src' },
  },
  {
    name: 'binary files match',
    left: () => {
      const v = new Volume()
      v.writeFileSync('/bin.dat', Buffer.from([0xde, 0xad, 0xbe, 0xef]))
      return v
    },
    right: () => {
      const v = new Volume()
      v.writeFileSync('/bin.dat', Buffer.from([0xde, 0xad, 0xbe, 0xef]))
      return v
    },
    pass: true,
  },
  {
    name: 'binary files mismatch',
    left: () => {
      const v = new Volume()
      v.writeFileSync('/bin.dat', Buffer.from([0xde, 0xad, 0xbe, 0xef]))
      return v
    },
    right: () => {
      const v = new Volume()
      v.writeFileSync('/bin.dat', Buffer.from([0xca, 0xfe, 0xba, 0xbe]))
      return v
    },
    pass: false,
  },
  {
    name: 'large binary file mismatch',
    left: () => {
      const v = new Volume()
      const buf = Buffer.alloc(100_000, 0xaa) // 100KB of 0xaa
      v.writeFileSync('/big.bin', buf)
      return v
    },
    right: () => {
      const v = new Volume()
      const buf = Buffer.alloc(100_000, 0xaa)
      buf[50_000] = 0xbb // flip a byte in the middle
      v.writeFileSync('/big.bin', buf)
      return v
    },
    pass: false,
  },
  {
    name: 'not: identical files',
    left: { '/foo.txt': 'hi' },
    right: { '/foo.txt': 'hi' },
    pass: false, // it should fail because `.not` negates a match
    not: true,
  },
  {
    name: 'not: content mismatch',
    left: { '/foo.txt': 'hello' },
    right: { '/foo.txt': 'world' },
    pass: true, // should pass because mismatch is expected with `.not`
    not: true,
  },
  {
    name: 'not: extra file',
    left: { '/foo.txt': 'hi', '/bar.txt': 'extra' },
    right: { '/foo.txt': 'hi' },
    pass: true, // passes because with `.not` we want them NOT to match
    not: true,
  },
])

describe('toMatchVolume()', () => {
  describe('unit', () => {
    const mockState = {
      utils: {
        printReceived: (received) => `received(${JSON.stringify(received)})`,
        matcherHint: (matcherName: string) => `hint(${matcherName})`,
      },
    }

    // invokes the matcher directly to get the return value snapshot
    async function testRunnerUnit({ left, right, opts, pass, not }: TestCase) {
      const leftVol = makeVol(left)
      const rightVol = makeVol(right)
      const invoke = async (fullReport = false) => {
        try {
          const matcher = toMatchVolume.bind(mockState as any)
          const result = await matcher(
            leftVol,
            rightVol,
            fullReport ? { ...opts, report: 'all' } : opts,
          )
          return { ...result, message: result.message() }
        } catch (e) {
          return e
        }
      }
      const result = await invoke()
      if (pass != null) expect(result).toHaveProperty('pass', not ? !pass : pass)
      expect(result).toMatchSnapshot('result')
      if (pass != null) {
        const result2 = await invoke(true)
        expect(result2).toHaveProperty('pass', not ? !pass : pass)
        expect(result2).toMatchSnapshot('result-full')
      }
    }

    it.each(cases.normal)('$name', testRunnerUnit)
    it.only.each(cases.only)('$name', testRunnerUnit)
    it.skip.each(cases.skip)('$name', testRunnerUnit)
  })

  describe('integration', () => {
    function testRunnerInteg({ left, right, pass, opts, not }: TestCase) {
      const leftVol = makeVol(left)
      const rightVol = makeVol(right)

      if (not) {
        if (pass) {
          expect(leftVol).not.toMatchVolume(rightVol, opts)
        } else {
          expect(() =>
            expect(leftVol).not.toMatchVolume(rightVol, opts),
          ).toThrowErrorMatchingSnapshot()
        }
      } else {
        if (pass) {
          expect(leftVol).toMatchVolume(rightVol, opts)
        } else {
          expect(() => expect(leftVol).toMatchVolume(rightVol, opts)).toThrowErrorMatchingSnapshot()
        }
      }
    }

    it.each(cases.normal)('$name', testRunnerInteg)
    it.only.each(cases.only)('$name', testRunnerInteg)
    it.skip.each(cases.skip)('$name', testRunnerInteg)
  })
})
