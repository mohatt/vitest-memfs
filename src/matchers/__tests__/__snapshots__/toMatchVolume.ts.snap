// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`toMatchVolume() > integration > 'binary files mismatch' 1`] = `[Error: mismatch in file \`/bin.dat\`]`;

exports[`toMatchVolume() > integration > 'content mismatch' 1`] = `[Error: mismatch in file \`/foo.txt\`]`;

exports[`toMatchVolume() > integration > 'empty dir vs missing dir' 1`] = `[Error: directory structure didnâ€™t match]`;

exports[`toMatchVolume() > integration > 'extra file' 1`] = `[Error: directory structure didnâ€™t match]`;

exports[`toMatchVolume() > integration > 'invalid object type (expected)' 1`] = `[TypeError: You must provide a memfs Volume instance or plain JSON object to [2mexpect([22m[31mreceived[39m[2m).[22mtoMatchVolume[2m([22m[32mexpected[39m[2m)[22m, not \`object\`.]`;

exports[`toMatchVolume() > integration > 'invalid type (expected)' 1`] = `[TypeError: You must provide a memfs Volume instance or plain JSON object to [2mexpect([22m[31mreceived[39m[2m).[22mtoMatchVolume[2m([22m[32mexpected[39m[2m)[22m, not \`string\`.]`;

exports[`toMatchVolume() > integration > 'invalid type (received)' 1`] = `[Error: expected [31m"invalid"[39m to be a memfs Volume instance]`;

exports[`toMatchVolume() > integration > 'large binary file mismatch' 1`] = `[Error: mismatch in file \`/big.bin\`]`;

exports[`toMatchVolume() > integration > 'missing file' 1`] = `[Error: directory structure didnâ€™t match]`;

exports[`toMatchVolume() > integration > 'not: identical files' 1`] = `[Error: volumes matched]`;

exports[`toMatchVolume() > integration > 'respects ignore-extra option (mismatch)' 1`] = `[Error: volume is missing 1 expected file]`;

exports[`toMatchVolume() > integration > 'respects ignore-missing option (mismatch)' 1`] = `[Error: volume has 1 unexpected file]`;

exports[`toMatchVolume() > integration > 'symlink target mismatch' 1`] = `[Error: symlink target mismatch at \`/link.txt\`]`;

exports[`toMatchVolume() > unit > 'accepts json input' > result 1`] = `
{
  "message": "volumes matched",
  "pass": true,
}
`;

exports[`toMatchVolume() > unit > 'accepts json input' > result-full 1`] = `
{
  "message": "volumes matched",
  "pass": true,
}
`;

exports[`toMatchVolume() > unit > 'binary files match' > result 1`] = `
{
  "message": "volumes matched",
  "pass": true,
}
`;

exports[`toMatchVolume() > unit > 'binary files match' > result-full 1`] = `
{
  "message": "volumes matched",
  "pass": true,
}
`;

exports[`toMatchVolume() > unit > 'binary files mismatch' > result 1`] = `
{
  "actual": BinaryFile {
    "hash": "d78f8bb992a56a597f6c7a1fb918bb78271367eb",
    "length": 4,
    "preview": "3q2+7w==...3q2+7w==",
  },
  "expected": BinaryFile {
    "hash": "1465dad126d32bf7aa5d3b95a5c7c331cdff34df",
    "length": 4,
    "preview": "yv66vg==...yv66vg==",
  },
  "message": "mismatch in file \`/bin.dat\`",
  "pass": false,
}
`;

exports[`toMatchVolume() > unit > 'binary files mismatch' > result-full 1`] = `
{
  "actual": {
    "/bin.dat": BinaryFile {
      "hash": "d78f8bb992a56a597f6c7a1fb918bb78271367eb",
      "length": 4,
      "preview": "3q2+7w==...3q2+7w==",
    },
  },
  "expected": {
    "/bin.dat": BinaryFile {
      "hash": "1465dad126d32bf7aa5d3b95a5c7c331cdff34df",
      "length": 4,
      "preview": "yv66vg==...yv66vg==",
    },
  },
  "message": "Found 1 mismatched content",
  "pass": false,
}
`;

exports[`toMatchVolume() > unit > 'content mismatch' > result 1`] = `
{
  "actual": TextFile {
    "data": "hello",
  },
  "expected": TextFile {
    "data": "world",
  },
  "message": "mismatch in file \`/foo.txt\`",
  "pass": false,
}
`;

exports[`toMatchVolume() > unit > 'content mismatch' > result-full 1`] = `
{
  "actual": {
    "/foo.txt": TextFile {
      "data": "hello",
    },
  },
  "expected": {
    "/foo.txt": TextFile {
      "data": "world",
    },
  },
  "message": "Found 1 mismatched content",
  "pass": false,
}
`;

exports[`toMatchVolume() > unit > 'empty dir vs missing dir' > result 1`] = `
{
  "actual": [
    "/empty",
  ],
  "expected": [
    "/",
  ],
  "message": "directory structure didnâ€™t match",
  "pass": false,
}
`;

exports[`toMatchVolume() > unit > 'empty dir vs missing dir' > result-full 1`] = `
{
  "actual": {
    "/empty": Directory {},
  },
  "expected": {
    "/": Directory {},
  },
  "message": "Found 2 mismatches: 1 missing path, 1 unexpected path",
  "pass": false,
}
`;

exports[`toMatchVolume() > unit > 'empty directories match' > result 1`] = `
{
  "message": "volumes matched",
  "pass": true,
}
`;

exports[`toMatchVolume() > unit > 'empty directories match' > result-full 1`] = `
{
  "message": "volumes matched",
  "pass": true,
}
`;

exports[`toMatchVolume() > unit > 'extra file' > result 1`] = `
{
  "actual": [
    "/bar.txt",
    "/foo.txt",
  ],
  "expected": [
    "/foo.txt",
  ],
  "message": "directory structure didnâ€™t match",
  "pass": false,
}
`;

exports[`toMatchVolume() > unit > 'extra file' > result-full 1`] = `
{
  "actual": {
    "/bar.txt": TextFile {
      "data": "extra",
    },
    "/foo.txt": {},
  },
  "expected": {
    "/foo.txt": {},
  },
  "message": "Found 1 unexpected path",
  "pass": false,
}
`;

exports[`toMatchVolume() > unit > 'identical files' > result 1`] = `
{
  "message": "volumes matched",
  "pass": true,
}
`;

exports[`toMatchVolume() > unit > 'identical files' > result-full 1`] = `
{
  "message": "volumes matched",
  "pass": true,
}
`;

exports[`toMatchVolume() > unit > 'invalid object type (expected)' > result 1`] = `[TypeError: You must provide a memfs Volume instance or plain JSON object to hint(toMatchVolume), not \`object\`.]`;

exports[`toMatchVolume() > unit > 'invalid type (expected)' > result 1`] = `[TypeError: You must provide a memfs Volume instance or plain JSON object to hint(toMatchVolume), not \`string\`.]`;

exports[`toMatchVolume() > unit > 'invalid type (received)' > result 1`] = `
{
  "actual": "invalid",
  "expected": Volume {},
  "message": "expected received("invalid") to be a memfs Volume instance",
  "pass": false,
}
`;

exports[`toMatchVolume() > unit > 'large binary file mismatch' > result 1`] = `
{
  "actual": BinaryFile {
    "hash": "cd561c0d1e53dbd0cd390f4b2cd3728db7dd71fc",
    "length": 100000,
    "preview": "qqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqo=...qqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqo=",
  },
  "expected": BinaryFile {
    "hash": "814bf2767d4339019bf26a4d3a35066ad9c0f6c7",
    "length": 100000,
    "preview": "qqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqo=...qqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqo=",
  },
  "message": "mismatch in file \`/big.bin\`",
  "pass": false,
}
`;

exports[`toMatchVolume() > unit > 'large binary file mismatch' > result-full 1`] = `
{
  "actual": {
    "/big.bin": BinaryFile {
      "hash": "cd561c0d1e53dbd0cd390f4b2cd3728db7dd71fc",
      "length": 100000,
      "preview": "qqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqo=...qqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqo=",
    },
  },
  "expected": {
    "/big.bin": BinaryFile {
      "hash": "814bf2767d4339019bf26a4d3a35066ad9c0f6c7",
      "length": 100000,
      "preview": "qqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqo=...qqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqo=",
    },
  },
  "message": "Found 1 mismatched content",
  "pass": false,
}
`;

exports[`toMatchVolume() > unit > 'missing file' > result 1`] = `
{
  "actual": [
    "/foo.txt",
  ],
  "expected": [
    "/bar.txt",
    "/foo.txt",
  ],
  "message": "directory structure didnâ€™t match",
  "pass": false,
}
`;

exports[`toMatchVolume() > unit > 'missing file' > result-full 1`] = `
{
  "actual": {
    "/foo.txt": {},
  },
  "expected": {
    "/bar.txt": TextFile {
      "data": "extra",
    },
    "/foo.txt": {},
  },
  "message": "Found 1 missing path",
  "pass": false,
}
`;

exports[`toMatchVolume() > unit > 'not: content mismatch' > result 1`] = `
{
  "actual": TextFile {
    "data": "hello",
  },
  "expected": TextFile {
    "data": "world",
  },
  "message": "mismatch in file \`/foo.txt\`",
  "pass": false,
}
`;

exports[`toMatchVolume() > unit > 'not: content mismatch' > result-full 1`] = `
{
  "actual": {
    "/foo.txt": TextFile {
      "data": "hello",
    },
  },
  "expected": {
    "/foo.txt": TextFile {
      "data": "world",
    },
  },
  "message": "Found 1 mismatched content",
  "pass": false,
}
`;

exports[`toMatchVolume() > unit > 'not: extra file' > result 1`] = `
{
  "actual": [
    "/bar.txt",
    "/foo.txt",
  ],
  "expected": [
    "/foo.txt",
  ],
  "message": "directory structure didnâ€™t match",
  "pass": false,
}
`;

exports[`toMatchVolume() > unit > 'not: extra file' > result-full 1`] = `
{
  "actual": {
    "/bar.txt": TextFile {
      "data": "extra",
    },
    "/foo.txt": {},
  },
  "expected": {
    "/foo.txt": {},
  },
  "message": "Found 1 unexpected path",
  "pass": false,
}
`;

exports[`toMatchVolume() > unit > 'not: identical files' > result 1`] = `
{
  "message": "volumes matched",
  "pass": true,
}
`;

exports[`toMatchVolume() > unit > 'not: identical files' > result-full 1`] = `
{
  "message": "volumes matched",
  "pass": true,
}
`;

exports[`toMatchVolume() > unit > 'respects ignore-extra option (mismatch)' > result 1`] = `
{
  "actual": [
    "/foo.txt",
  ],
  "expected": [
    "/bar.txt",
    "/foo.txt",
  ],
  "message": "volume is missing 1 expected file",
  "pass": false,
}
`;

exports[`toMatchVolume() > unit > 'respects ignore-extra option (mismatch)' > result-full 1`] = `
{
  "actual": {
    "/foo.txt": {},
  },
  "expected": {
    "/bar.txt": TextFile {
      "data": "hey",
    },
    "/foo.txt": {},
  },
  "message": "Found 1 missing path",
  "pass": false,
}
`;

exports[`toMatchVolume() > unit > 'respects ignore-extra option' > result 1`] = `
{
  "message": "volumes matched",
  "pass": true,
}
`;

exports[`toMatchVolume() > unit > 'respects ignore-extra option' > result-full 1`] = `
{
  "message": "volumes matched",
  "pass": true,
}
`;

exports[`toMatchVolume() > unit > 'respects ignore-missing option (mismatch)' > result 1`] = `
{
  "actual": [
    "/extra.txt",
    "/foo.txt",
  ],
  "expected": [
    "/foo.txt",
  ],
  "message": "volume has 1 unexpected file",
  "pass": false,
}
`;

exports[`toMatchVolume() > unit > 'respects ignore-missing option (mismatch)' > result-full 1`] = `
{
  "actual": {
    "/extra.txt": TextFile {
      "data": "extra",
    },
    "/foo.txt": {},
  },
  "expected": {
    "/foo.txt": {},
  },
  "message": "Found 1 unexpected path",
  "pass": false,
}
`;

exports[`toMatchVolume() > unit > 'respects ignore-missing option' > result 1`] = `
{
  "message": "volumes matched",
  "pass": true,
}
`;

exports[`toMatchVolume() > unit > 'respects ignore-missing option' > result-full 1`] = `
{
  "message": "volumes matched",
  "pass": true,
}
`;

exports[`toMatchVolume() > unit > 'respects prefix option' > result 1`] = `
{
  "message": "volumes matched",
  "pass": true,
}
`;

exports[`toMatchVolume() > unit > 'respects prefix option' > result-full 1`] = `
{
  "message": "volumes matched",
  "pass": true,
}
`;

exports[`toMatchVolume() > unit > 'same ref (singleton vol)' > result 1`] = `
{
  "message": "volumes matched by reference",
  "pass": true,
}
`;

exports[`toMatchVolume() > unit > 'same ref (singleton vol)' > result-full 1`] = `
{
  "message": "volumes matched by reference",
  "pass": true,
}
`;

exports[`toMatchVolume() > unit > 'symlink target mismatch' > result 1`] = `
{
  "actual": Symlink {
    "target": "/target1.txt",
  },
  "expected": Symlink {
    "target": "/target2.txt",
  },
  "message": "symlink target mismatch at \`/link.txt\`",
  "pass": false,
}
`;

exports[`toMatchVolume() > unit > 'symlink target mismatch' > result-full 1`] = `
{
  "actual": {
    "/link.txt": Symlink {
      "target": "/target1.txt",
    },
  },
  "expected": {
    "/link.txt": Symlink {
      "target": "/target2.txt",
    },
  },
  "message": "Found 1 mismatched content",
  "pass": false,
}
`;
